File: app/src/main/java/com/example/zhpan/circleviewpager/adapter/AdapterFragmentPager.java
package com.example.zhpan.circleviewpager.adapter;

import android.util.SparseArray;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;


import com.example.zhpan.circleviewpager.fragment.BaseFragment;
import com.example.zhpan.circleviewpager.fragment.IndicatorFragment;
import com.example.zhpan.circleviewpager.fragment.PageFragment;
import com.example.zhpan.circleviewpager.fragment.HomeFragment;

public class AdapterFragmentPager extends FragmentPagerAdapter {

    public static final int PAGE_HOME = 0;

    public static final int PAGE_FIND = 1;

    public static final int PAGE_OTHERS = 2;

    private SparseArray<BaseFragment> fragmentList;

    public AdapterFragmentPager(FragmentManager fm) {
        super(fm);
        fragmentList = getFragments();
    }

    private SparseArray<BaseFragment> getFragments() {
        SparseArray<BaseFragment> fragmentList = new SparseArray<>();
        fragmentList.put(PAGE_HOME, HomeFragment.getInstance());
        fragmentList.put(PAGE_FIND, PageFragment.getInstance());
        fragmentList.put(PAGE_OTHERS, IndicatorFragment.getInstance());
        return fragmentList;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        Fragment fragment = (Fragment) super.instantiateItem(container, position);
        fragmentList.put(position, (BaseFragment) fragment);
        return fragment;
    }

    @Override
    public Fragment getItem(int position) {
        return fragmentList.get(position);
    }

    @Override
    public int getCount() {
        if (fragmentList != null) {
            return fragmentList.size();
        } else {
            return 0;
        }
    }

}

File: app/src/main/java/com/example/zhpan/circleviewpager/adapter/AdapterFragmentPager.kt
package com.example.zhpan.circleviewpager.adapter

import android.util.SparseArray
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter

import com.example.zhpan.circleviewpager.fragment.BaseFragment
import com.example.zhpan.circleviewpager.fragment.HomeFragment
import com.example.zhpan.circleviewpager.fragment.IndicatorFragment
import com.example.zhpan.circleviewpager.fragment.OthersFragment
import com.example.zhpan.circleviewpager.fragment.PageFragment

/**
 * <pre>
 * Created by zhangpan on 2019-12-05.
 * Description:
</pre> *
 */
class AdapterFragmentPager(fragmentActivity: FragmentActivity) : FragmentStateAdapter(fragmentActivity) {

    private val fragments: SparseArray<BaseFragment> = SparseArray()

    override fun createFragment(position: Int): Fragment {
        var fragment: Fragment
        when (position) {
            PAGE_HOME -> {
                if (fragments.get(PAGE_HOME) == null) {
                    fragment = HomeFragment.getInstance();
                    fragments.put(PAGE_HOME, fragment)
                } else {
                    fragment = fragments.get(PAGE_HOME)
                }
            }
            PAGE_FIND -> {
                if (fragments.get(PAGE_FIND) == null) {
                    fragment = PageFragment.getInstance();
                    fragments.put(PAGE_FIND, fragment)
                } else {
                    fragment = fragments.get(PAGE_FIND)
                }
            }

            PAGE_INDICATOR -> {
                if (fragments.get(PAGE_INDICATOR) == null) {
                    fragment = IndicatorFragment.getInstance();
                    fragments.put(PAGE_INDICATOR, fragment)
                } else {
                    fragment = fragments.get(PAGE_INDICATOR)
                }
            }

            PAGE_OTHERS -> {
                if (fragments.get(PAGE_OTHERS) == null) {
                    fragment = OthersFragment.getInstance();
                    fragments.put(PAGE_OTHERS, fragment)
                } else {
                    fragment = fragments.get(PAGE_OTHERS)
                }
            }
            else -> {
                if (fragments.get(PAGE_HOME) == null) {
                    fragment = HomeFragment.getInstance();
                    fragments.put(PAGE_HOME, fragment)
                } else {
                    fragment = fragments.get(PAGE_HOME)
                }
            }
        }
        return fragment
    }

    override fun getItemCount(): Int {
        return 4
    }

    companion object {

        const val PAGE_HOME = 0

        const val PAGE_FIND = 1

        const val PAGE_INDICATOR = 2

        const val PAGE_OTHERS = 3

    }

}
